File: packages/governance/src/outcome.ts
import { createProof } from '@totemsdk/proof'
import type { SignedProof, UnsignedProof } from '@totemsdk/proof'
import type { MandateBody } from '@totemsdk/authority'
import type { Proposal, ProposalOutcome, ProposalAction, VoteTally } from './types.js'
import { computeTallyProofHash } from './tally.js'

export function createOutcome(params: {
  proposal: Proposal
  tally: VoteTally
  determinedBy: string
  determinedAt?: number
}): ProposalOutcome {
  const tallyHash = computeTallyProofHash(params.tally)
  const now = params.determinedAt ?? Date.now()

  return {
    proposalId: params.proposal.id,
    status: params.tally.passed ? 'passed' : 'failed',
    tallyHash,
    passed: params.tally.passed,
    determinedAt: now,
    determinedBy: params.determinedBy,
  }
}

export function createOutcomeProofDraft(
  outcome: ProposalOutcome,
  issuer: string,
): UnsignedProof {
  return createProof({
    kind: 'custom',
    subject: { id: outcome.proposalId, kind: 'proposal-outcome' },
    issuer,
    issuedAt: outcome.determinedAt,
    payload: {
      schema: 'totem:governance:outcome/v1',
      outcome: {
        proposalId: outcome.proposalId,
        status: outcome.status,
        tallyHash: outcome.tallyHash,
        passed: outcome.passed,
      },
    },
  })
}

export function createGovernedMandate(
  outcome: ProposalOutcome,
  action: ProposalAction,
  actionIndex: number,
  governanceIdentity: string,
  executor: string,
  params: {
    membershipSnapshotHash: string
    voteTallyHash: string
    outcomeProofId: string
  },
): MandateBody {
  const constraints: MandateBody['constraints'] = [
    { field: 'proposalId', operator: 'eq', value: outcome.proposalId },
    { field: 'actionIndex', operator: 'eq', value: actionIndex },
    { field: 'actionType', operator: 'eq', value: action.type },
    { field: 'membershipSnapshotHash', operator: 'eq', value: params.membershipSnapshotHash },
    { field: 'voteTallyHash', operator: 'eq', value: params.voteTallyHash },
    { field: 'outcomeProofId', operator: 'eq', value: params.outcomeProofId },
  ]

  if (action.target) {
    constraints.push({ field: 'target', operator: 'eq', value: action.target })
  }

  const actionPayload = action.payload
  const payloadKeys = Object.keys(actionPayload).sort()
  for (const key of payloadKeys) {
    constraints.push({
      field: `payload.${key}`,
      operator: 'eq',
      value: actionPayload[key],
    })
  }

  return {
    grantor: governanceIdentity,
    grantee: executor,
    principal: outcome.proposalId,
    scope: `governance:${action.type}:execute`,
    constraints,
    usageLimit: { maxCount: 1 },
    issuedAt: outcome.determinedAt,
  }
}

export function finalizeProposalExecution(
  proposal: Proposal,
  outcomeProofSigned: SignedProof,
  txId: string,
): Proposal {
  return {
    ...proposal,
    status: 'executed',
    executedAt: Date.now(),
    executionTxId: txId,
  }
}

File: packages/governance/rust/src/outcome.rs
use serde::Deserialize;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::tally::compute_tally_proof_hash;
use crate::types::{
    GovernedMandate, MandateConstraint, Proposal, ProposalAction, ProposalOutcome, ProposalStatus,
    UsageLimit, VoteTally,
};

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

#[derive(Deserialize)]
pub struct CreateOutcomeParams {
    pub proposal: Proposal,
    pub tally: VoteTally,
    pub determined_by: String,
    pub determined_at: Option<u64>,
}

pub fn create_outcome(params: CreateOutcomeParams) -> ProposalOutcome {
    let tally_hash = compute_tally_proof_hash(&params.tally);
    let now = params.determined_at.unwrap_or_else(now_ms);

    ProposalOutcome {
        proposal_id: params.proposal.id,
        status: if params.tally.passed {
            ProposalStatus::Passed
        } else {
            ProposalStatus::Failed
        },
        tally_hash,
        passed: params.tally.passed,
        determined_at: now,
        determined_by: params.determined_by,
        proof_id: None,
    }
}

#[derive(Deserialize)]
pub struct CreateGovernedMandateParams {
    pub outcome: ProposalOutcome,
    pub action: ProposalAction,
    pub action_index: u32,
    pub governance_identity: String,
    pub executor: String,
    pub membership_snapshot_hash: String,
    pub vote_tally_hash: String,
    pub outcome_proof_id: String,
}

pub fn create_governed_mandate(params: CreateGovernedMandateParams) -> GovernedMandate {
    let mut constraints: Vec<MandateConstraint> = vec![
        MandateConstraint {
            field: "proposalId".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::String(params.outcome.proposal_id.clone()),
        },
        MandateConstraint {
            field: "actionIndex".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::Number(serde_json::Number::from(params.action_index)),
        },
        MandateConstraint {
            field: "actionType".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::String(format!("{:?}", params.action.action_type)),
        },
        MandateConstraint {
            field: "membershipSnapshotHash".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::String(params.membership_snapshot_hash),
        },
        MandateConstraint {
            field: "voteTallyHash".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::String(params.vote_tally_hash),
        },
        MandateConstraint {
            field: "outcomeProofId".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::String(params.outcome_proof_id),
        },
    ];

    if let Some(ref target) = params.action.target {
        constraints.push(MandateConstraint {
            field: "target".to_string(),
            operator: "eq".to_string(),
            value: serde_json::Value::String(target.clone()),
        });
    }

    if let serde_json::Value::Object(ref payload_map) = params.action.payload {
        let mut keys: Vec<&String> = payload_map.keys().collect();
        keys.sort();
        for key in keys {
            constraints.push(MandateConstraint {
                field: format!("payload.{}", key),
                operator: "eq".to_string(),
                value: payload_map[key].clone(),
            });
        }
    }

    GovernedMandate {
        grantor: params.governance_identity,
        grantee: params.executor,
        principal: params.outcome.proposal_id,
        scope: format!("governance:{:?}:execute", params.action.action_type),
        constraints,
        usage_limit: UsageLimit { max_count: 1 },
        issued_at: params.outcome.determined_at,
    }
}

pub fn finalize_proposal_execution(
    proposal: &Proposal,
    _outcome_proof_signed: &serde_json::Value,
    tx_id: &str,
) -> Proposal {
    let mut p = proposal.clone();
    p.status = ProposalStatus::Executed;
    p.executed_at = Some(now_ms());
    p.execution_tx_id = Some(tx_id.to_string());
    p
}
